components/edit-collective/sections/SetupTwoFactorAuth.js

```diff
@@ -0,0 +1,242 @@
+import React, { Fragment } from 'react';
+import PropTypes from 'prop-types';
+import { graphql } from '@apollo/react-hoc';
+import { CheckCircle } from '@styled-icons/fa-solid/CheckCircle';
+import { Field, Form, Formik } from 'formik';
+import QRCode from 'qrcode.react';
+import { defineMessages, FormattedMessage, injectIntl } from 'react-intl';
+import speakeasy from 'speakeasy';
+
+import { getErrorFromGraphqlException } from '../../../lib/errors';
+import { API_V2_CONTEXT, gqlV2 } from '../../../lib/graphql/helpers';
+import { compose } from '../../../lib/utils';
+
+import Container from '../../Container';
+import { Box, Flex } from '../../Grid';
+import MessageBox from '../../MessageBox';
+import StyledButton from '../../StyledButton';
+import StyledInput from '../../StyledInput';
+import StyledInputField from '../../StyledInputField';
+import { H2, H3, P } from '../../Text';
+import { withUser } from '../../UserProvider';
+
+const messages = defineMessages({
+  errorWrongLength: {
+    id: 'TwoFactorAuth.Setup.Form.ErrorLength',
+    defaultMessage: 'Incorrect code length. Please re-enter your 6-digit code.',
+  },
+  errorWrongFormat: {
+    id: 'TwoFactorAuth.Setup.Form.ErrorFormat',
+    defaultMessage: 'Incorrect code format. Please enter only numbers.',
+  },
+  inputLabel: {
+    id: 'TwoFactorAuth.Setup.Form.InputLabel',
+    defaultMessage: 'Please enter your 6-digit code without any dashes.',
+  },
+});
+
+class SetupTwoFactorAuth extends React.Component {
+  static propTypes = {
+    collective: PropTypes.string.isRequired,
+    /** From intl */
+    intl: PropTypes.object.isRequired,
+    /** From graphql query */
+    addTwoFactorAuthToAccount: PropTypes.func.isRequired,
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      error: null,
+      tokenAdded: null,
+    };
+  }
+
+  /* TO DO
+
+  - Check the Collective gql info and see if there is already 2FA, if so then we display a different section
+  - Error handling for if the secret gets rejected
+  - Need to figure out how to change the state without everything reloading and redoing the QR code :-/
+
+  */
+
+  render() {
+    const { intl, collective } = this.props;
+    const { error, tokenAdded } = this.state;
+
+    const doesAccountAlreadyHave2FA = tokenAdded || collective.hasTwoFactorAuth;
+
+    const initialValues = {
+      TOTPCode: '',
+    };
+
+    const validate = values => {
+      const errors = {};
+
+      if (values.TOTPCode.toString().length > 6) {
+        errors.TOTPCode = intl.formatMessage(messages.errorWrongLength);
+      }
+
+      return errors;
+    };
+
+    const options = {
+      issuer: 'Open Collective',
+      name: `${collective.name}'s Open Collective`,
+      length: 64,
+    };
+
+    const secret = speakeasy.generateSecret(options);
```

Comment:
We'll have to review security implications of this.

Also for bundling reasons, maybe we want to generate that secret server side, so we don't have to bundle the speakeasy library (not sure how heavy it is).